plugins/completion/nvim-cmp/default.nix

```diff
@@ -115,103 +79,158 @@ in {
       '';
     };
 
-    completion = mkOption {
-      default = null;
-      type = types.nullOr (types.submodule ({...}: {
-        options = {
-          keyword_length = mkOption {
-            default = null;
-            type = types.nullOr types.int;
-          };
+    mappingPresets = mkOption {
+      default = [];
+      type = types.listOf (types.enum [
+        "insert"
+        "cmdline"
+        # Not sure if there are more or if this should just be str
+      ]);
+      description = ''
+        Mapping presets to use; cmp.mapping.preset.
+        \$\{mappingPreset} will be called with the configured mappings.
+      '';
+      example = ''[ "insert" "cmdline" ]'';
+    };
 
-          keyword_pattern = mkOption {
-            default = null;
-            type = types.nullOr types.str;
-          };
+    snippet = helpers.mkCompositeOption "Snippet options" {
+      expand =
+        helpers.defaultNullOpts.mkStr
+        ''
+          function(_)
+            error('snippet engine is not configured.')
+          end
+        ''
+        ''
+          The snippet expansion function. That's how nvim-cmp interacts with a
+          particular snippet engine.
+
+          Example:
+          ```
+          function(args)
+            vim.fn["vsnip#anonymous"](args.body) -- For `vsnip` users.
+            -- require('luasnip').lsp_expand(args.body) -- For `luasnip` users.
+            -- require('snippy').expand_snippet(args.body) -- For `snippy` users.
+            -- vim.fn["UltiSnips#Anon"](args.body) -- For `ultisnips` users.
+          end
+          ```
+        '';
+    };
```

Comment:
I feel the problem with this one is  that we don't really know which engine is selected so we can't install it. Maybe in a future commit we could be a bit more intelligent with an enum for the diffrent possible engines, and still leave an escape hatch to plug your own ?

Reply:
Ok, I'll do it in another PR.